File: src/LanAudioRelay/Services/SenderSession.cs
using System.Net;
using System.Net.Sockets;
using LanAudioRelay.Audio;
using LanAudioRelay.Core.Audio;
using LanAudioRelay.Core.Networking;
using LanAudioRelay.Core.Protocol;

namespace LanAudioRelay.Services;

public sealed class SenderSession : IAsyncDisposable
{
    private readonly object _gate = new();
    private readonly ControlClient _controlClient = new();
    private readonly WasapiLoopbackAudioSource _audioSource = new();
    private readonly OpusFrameEncoder _encoder = new();
    private UdpClient? _udp;
    private IPEndPoint? _mediaEndPoint;
    private Guid _sessionId;
    private uint _sequence;
    private ulong _timestamp;
    private bool _running;

    public SenderSession()
    {
        _audioSource.FrameCaptured += OnFrameCaptured;
        _audioSource.Faulted += (_, message) => StatusChanged?.Invoke(this, $"Capture error: {message}");
    }

    public event EventHandler<string>? StatusChanged;
    public event EventHandler<float>? InputLevelChanged;

    public async Task StartAsync(string receiverHost, string pairingCode, CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            if (_running)
            {
                throw new InvalidOperationException("Sender is already running.");
            }

            _running = true;
        }

        try
        {
            StatusChanged?.Invoke(this, "Pairing with receiver...");
            var session = await _controlClient
                .ConnectAndPairAsync(receiverHost, pairingCode, cancellationToken: cancellationToken)
                .ConfigureAwait(false);

            _udp = new UdpClient(AddressFamily.InterNetwork);
            _mediaEndPoint = new IPEndPoint(IPAddress.Parse(session.ReceiverHost), session.MediaPort);
            _sessionId = session.SessionId;
            _sequence = 0;
            _timestamp = 0;

            _audioSource.Start();
            StatusChanged?.Invoke(this, $"Streaming to {receiverHost}:{session.MediaPort}");
        }
        catch
        {
            await StopAsync().ConfigureAwait(false);
            throw;
        }
    }

    private void OnFrameCaptured(object? sender, PcmFrameCapturedEventArgs e)
    {
        UdpClient? udp;
        IPEndPoint? endpoint;
        Guid sessionId;
        uint sequence;
        ulong timestamp;

        lock (_gate)
        {
            if (!_running || _udp is null || _mediaEndPoint is null)
            {
                return;
            }

            udp = _udp;
            endpoint = _mediaEndPoint;
            sessionId = _sessionId;
            sequence = _sequence++;
            timestamp = _timestamp;
            _timestamp += AudioConstants.SamplesPerFrame;
        }

        try
        {
            var payload = _encoder.Encode(e.Pcm);
            var packet = new AudioPacket(sessionId, sequence, timestamp, AudioConstants.CodecOpus, payload);
            var bytes = packet.ToArray();
            udp.Send(bytes, bytes.Length, endpoint);
            InputLevelChanged?.Invoke(this, e.Level);
        }
        catch (ObjectDisposedException)
        {
        }
        catch (SocketException ex)
        {
            StatusChanged?.Invoke(this, $"UDP send error: {ex.Message}");
        }
    }

    public async Task StopAsync()
    {
        lock (_gate)
        {
            _running = false;
        }

        _audioSource.Stop();
        _udp?.Dispose();
        _udp = null;
        _mediaEndPoint = null;
        StatusChanged?.Invoke(this, "Sender stopped.");
        await Task.CompletedTask.ConfigureAwait(false);
    }

    public async ValueTask DisposeAsync()
    {
        await StopAsync().ConfigureAwait(false);
        _audioSource.Dispose();
    }
}

File: mac/LanAudioRelayMac/Sources/LanAudioRelayMac/SenderSession.swift
import Foundation
import LanAudioRelayMacCore

final class SenderSession {
    private let controlClient = ControlClient()
    private let capture = MacSystemAudioCapture()
    private let lock = NSLock()
    private var encoder: OpusFrameEncoder?
    private var mediaSender: MediaSender?
    private var sessionId: UUID?
    private var sequence: UInt32 = 0
    private var timestamp: UInt64 = 0
    private var running = false

    var onStatus: (@Sendable (String) -> Void)?
    var onInputLevel: (@Sendable (Float) -> Void)?

    init() {
        capture.onFrame = { [weak self] frame in
            self?.send(frame)
        }
        capture.onError = { [weak self] message in
            self?.onStatus?("Capture error: \(message)")
        }
    }

    func start(receiverHost: String, pairingCode: String) async throws {
        lock.lock()
        if running {
            lock.unlock()
            throw NSError(domain: "LanAudioRelayMac", code: 1, userInfo: [NSLocalizedDescriptionKey: "Sender is already running."])
        }
        running = true
        lock.unlock()

        do {
            onStatus?("Pairing with receiver...")
            let session = try await controlClient.connectAndPair(host: receiverHost, pairingCode: pairingCode)
            let encoder = try OpusFrameEncoder()
            let sender = try MediaSender(host: session.receiverHost, port: UInt16(session.mediaPort))

            lock.lock()
            self.encoder = encoder
            self.mediaSender = sender
            self.sessionId = session.sessionId
            self.sequence = 0
            self.timestamp = 0
            lock.unlock()

            try await capture.start()
            onStatus?("Streaming to \(receiverHost):\(session.mediaPort)")
        } catch {
            await stop()
            throw error
        }
    }

    private func send(_ frame: PCMFrame) {
        lock.lock()
        guard running,
              let encoder,
              let mediaSender,
              let sessionId
        else {
            lock.unlock()
            return
        }

        let sequence = self.sequence
        let timestamp = self.timestamp
        self.sequence &+= 1
        self.timestamp &+= UInt64(AudioConstants.samplesPerFrame)
        lock.unlock()

        do {
            let payload = try encoder.encode(frame.samples)
            let packet = AudioPacket(
                sessionId: sessionId,
                sequence: sequence,
                timestamp: timestamp,
                codec: AudioConstants.codecOpus,
                payload: payload
            )
            try mediaSender.send(packet)
            onInputLevel?(frame.level)
        } catch {
            onStatus?("Send error: \(error.localizedDescription)")
        }
    }

    func stop() async {
        lock.lock()
        running = false
        encoder = nil
        mediaSender = nil
        sessionId = nil
        lock.unlock()

        await capture.stop()
        onInputLevel?(0)
        onStatus?("Sender stopped.")
    }
}
